File: companion/lib/gitStats.js
const { execFile } = require('child_process');

function parseGitShortstat(text) {
  const s = String(text || '');
  const ins = s.match(/(\d+) insertion/);
  const del = s.match(/(\d+) deletion/);
  return { added: ins ? Number(ins[1]) : 0, modified: del ? Number(del[1]) : 0 };
}

function git(cwd, args, timeout = 4000) {
  return new Promise((resolve) => {
    execFile('git', ['-C', cwd, ...args], { timeout, maxBuffer: 4 * 1024 * 1024 },
      (err, stdout) => resolve(err ? '' : stdout.trim()));
  });
}

async function readGitStats(cwd) {
  const branch = await git(cwd, ['rev-parse', '--abbrev-ref', 'HEAD']);
  const stat = await git(cwd, ['diff', '--shortstat']);
  return { branch: branch || '', ...parseGitShortstat(stat) };
}

module.exports = { parseGitShortstat, readGitStats };

File: companion/codey/gitstats.py
"""git diff --shortstat 解析 + 读取(本地子进程,无网络)。"""
import re
import subprocess


def parse_git_shortstat(text):
    s = str(text if text is not None else "")
    ins = re.search(r"(\d+) insertion", s)
    dele = re.search(r"(\d+) deletion", s)
    return {
        "added": int(ins.group(1)) if ins else 0,
        "modified": int(dele.group(1)) if dele else 0,
    }


def _git(cwd, args, timeout=4):
    try:
        r = subprocess.run(["git", "-C", cwd, *args],
                           capture_output=True, text=True, timeout=timeout)
        return r.stdout.strip() if r.returncode == 0 else ""
    except Exception:
        return ""


def read_git_stats(cwd):
    if not cwd:
        return {"branch": "", "added": 0, "modified": 0}
    branch = _git(cwd, ["rev-parse", "--abbrev-ref", "HEAD"])
    stat = _git(cwd, ["diff", "--shortstat"])
    return {"branch": branch or "", **parse_git_shortstat(stat)}
